### PR TITLE
135 split package cr

### PR DESCRIPTION
**Pull request by ewrenn8**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -59,8 +59,7 @@ RUN wget -O- https://github.com/mozilla/sops/releases/download/v3.6.1/sops-v3.6.
   chmod +x /usr/local/bin/sops && sops -v
 
 # kapp-controller
-COPY . .
-RUN CGO_ENABLED=0 GOOS=linux go build -mod=vendor -ldflags=-buildid= -trimpath -o controller ./cmd/main.go
+COPY ./controller ./controller
```
Comment (cppforlife) [inline Dockerfile]: revert this change?

File: cmd/controller/handlers/installed_pkg_version_handler.go [hunk 2]
```diff
@@ -87,14 +87,7 @@ func (ipvh *InstalledPkgVersionHandler) enqueueEligibleInstalledPackages(q workq
...
 func (ipvh *InstalledPkgVersionHandler) isEligibleForVersionUpgrade(version string, installedPkg ipkgv1alpha1.InstalledPackage) bool {
-	semverConfig := installedPkg.Spec.PkgRef.VersionSelection
-	if installedPkg.Spec.PkgRef.Version != "" {
-		semverConfig = &versions.VersionSelectionSemver{
-			Constraints: installedPkg.Spec.PkgRef.Version,
-			// Prereleases must be non nil to be included
-			Prereleases: &versions.VersionSelectionSemverPrereleases{},
-		}
-	}
+	semverConfig := installedPkg.Spec.VersionSelection
```
Comment (cppforlife) [inline cmd/controller/handlers/installed_pkg_version_handler.go]: are we not maintaining single version field?

File: hack/gen-apiserver.sh [hunk 3]
```diff
@@ -9,49 +9,49 @@ GOPATH=`go env GOPATH`
...
-  --output-package "${KC_PKG}/pkg/apiserver/client/informers" \
-  --go-header-file hack/gen-boilerplate.txt
-
-echo "Generating deepcopy"
-$GOPATH/bin/deepcopy-gen \
-  --input-dirs "${KC_PKG}/pkg/apiserver/apis/packages/v1alpha1" \
-  --input-dirs "${KC_PKG}/pkg/apiserver/apis/packages" \
-  -O zz_generated.deepcopy \
-  --go-header-file hack/gen-boilerplate.txt
+# echo "Generating clients"
```
Comment (cppforlife) [inline hack/gen-apiserver.sh]: removed?

File: pkg/apis/installpackage/v1alpha1/installed_package.go [hunk 4]
```diff
@@ -46,22 +46,15 @@ type InstalledPackageSpec struct {
...
+	PackageName string `json:"packageName,omitempty"`
+	// +optional
+	VersionSelection *versions.VersionSelectionSemver `json:"versionSelection,omitempty"`
 	// +optional
 	Values []InstalledPackageValues `json:"values,omitempty"`
 
 	// TODO other App CR related fields
 }
 
-type PackageRef struct {
```
Comment (cppforlife) [inline pkg/apis/installpackage/v1alpha1/installed_package.go:71]: hmm. im going back and forth on should we have

```
spec:
  packageVersionRef:
    packageName: ...
    version: ...
    versionSelection: ...
```

File: test.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
+---
```
Comment (cppforlife) [inline test.yml]: delete this file?

Comment (cppforlife) [inline test.yml:1]: delete this file?

File: pkg/apiserver/apis/packages/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	Version                         string   `json:"version,omitempty"`
+	Licenses                        []string `json:"licenses,omitempty"`
+	ReleasedAt                      string   `json:"releasedAt,omitempty"`
+	CapactiyRequirementsDescription string   `json:"capacityRequirementsDescription,omitempty"`
+	ReleaseNotes                    string   `json:"releaseNotes,omitempty"`
+
+	Template AppTemplateSpec `json:"template,omitempty"`
+
+	// valuesSchema can be used to show template values that
+	// can be configured by users when a Package is installed
```
Comment (danielhelfand) [inline pkg/apiserver/apis/packages/v1alpha1/types.go]: nit:

```suggestion
	// can be configured by users when a PackageVersion is installed
```

File: pkg/installedpkg/installedpkg_test.go [hunk 7]
```diff
@@ -88,21 +92,21 @@ func Test_PackageRefUsesName(t *testing.T) {
...
-	out, err := ip.referencedPkg()
+	out, err := ip.referencedPkgVersion()
 	if err != nil {
 		t.Fatalf("\nExpected no error from resolving referenced package\nBut got:\n%v\n", err)
 	}
 
-	if !reflect.DeepEqual(out, expectedPkg) {
-		t.Fatalf("\nPackage is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPkg, out)
+	if !reflect.DeepEqual(out, expectedPackageVersion) {
+		t.Fatalf("\nPackage is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPackageVersion, out)
```
Comment (danielhelfand) [inline pkg/installedpkg/installedpkg_test.go]:
```suggestion
		t.Fatalf("\nPackageVersion is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPackageVersion, out)
```

File: pkg/installedpkg/installedpkg_test.go [hunk 8]
```diff
@@ -37,49 +38,52 @@ func Test_PackageRefWithPrerelease_IsFound(t *testing.T) {
...
-	out, err := ip.referencedPkg()
+	out, err := ip.referencedPkgVersion()
 	if err != nil {
 		t.Fatalf("\nExpected no error from getting PackageRef with prerelease\nBut got:\n%v\n", err)
 	}
 
-	if !reflect.DeepEqual(out, expectedPkg) {
-		t.Fatalf("\nPackage is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPkg, out)
+	if !reflect.DeepEqual(out, expectedPackageVersion) {
+		t.Fatalf("\nPackage is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPackageVersion, out)
```
Comment (danielhelfand) [inline pkg/installedpkg/installedpkg_test.go]:
```suggestion
		t.Fatalf("\nPackageVersion is not same:\nExpected:\n%#v\nGot:\n%#v\n", expectedPackageVersion, out)
```
